<a href="https://colab.research.google.com/github/solomeya12345-code/python_for_data_analitic/blob/main/Homework_3_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Домашнє завдання до теми "Розширені методи обробки даних у Pandas"


В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [2]:

from google.colab import drive
drive.mount('/content/drive')


Mounted at /content/drive


In [3]:
import pandas as pd

In [4]:
df = pd.read_csv('drive/MyDrive/Python_for_DA/supermarket_sales.csv')

1. Відсортуйте дані за спаданням ціни товара ('Unit price') і виведіть 5 перших рядочків на екран.

In [5]:
df.sort_values(by="Unit price", ascending=False).head(5)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
983,148-41-7930,C,Naypyitaw,Normal,Male,Health and beauty,99.96,7,34.986,734.706,1/23/2019,10:33,Cash,699.72,4.761905,34.986,6.1
122,219-22-9386,B,Mandalay,Member,Male,Sports and travel,99.96,9,44.982,944.622,3/9/2019,17:26,Credit card,899.64,4.761905,44.982,4.2
930,641-62-7288,B,Mandalay,Normal,Male,Home and lifestyle,99.92,6,29.976,629.496,3/24/2019,13:33,Ewallet,599.52,4.761905,29.976,7.1
494,437-53-3084,B,Mandalay,Normal,Male,Fashion accessories,99.89,2,9.989,209.769,2/26/2019,11:48,Ewallet,199.78,4.761905,9.989,7.1
283,667-92-0055,A,Yangon,Member,Male,Health and beauty,99.83,6,29.949,628.929,3/4/2019,15:02,Ewallet,598.98,4.761905,29.949,8.5


2. Знайдіть інвойс (рядок в даних) в категорії товарів `Electronic accessories`, в якому був куплений товар з найбільшою кількістю і найбільшою ціною. Виведіть цей рядок даних на екран.

In [6]:
df[['Quantity', 'Unit price']].agg(['max'])

,Quantity,Unit price
max,10,99.96


In [7]:
df[['Quantity', 'Unit price']].max()

,0
Quantity,10.00
Unit price,99.96


In [8]:
subset = df[df['Product line'] == 'Electronic accessories']

raw_maxes = subset.sort_values(by=['Quantity', 'Unit price'], ascending=[False, False]).head(1)

display(raw_maxes)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
105,704-48-3927,A,Yangon,Member,Male,Electronic accessories,88.67,10,44.335,931.035,1/12/2019,14:50,Ewallet,886.7,4.761905,44.335,7.3


Розв'язок:

In [9]:
df[df['Product line']=='Electronic accessories'].sort_values(by=["Quantity", "Unit price"], ascending=[False, False]).head(1)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
105,704-48-3927,A,Yangon,Member,Male,Electronic accessories,88.67,10,44.335,931.035,1/12/2019,14:50,Ewallet,886.7,4.761905,44.335,7.3


In [10]:
df[df['Product line']=='Electronic accessories'].sort_values(by=['Unit price', 'Quantity'], ascending=[False, False])[:1]

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
209,817-69-8206,B,Mandalay,Normal,Female,Electronic accessories,99.73,9,44.8785,942.4485,3/2/2019,19:42,Credit card,897.57,4.761905,44.8785,6.5


3.1. Створіть нову колонку в даних `Rating Category`, яка визначає категорію рейтингу (`Rating`): високий, середній чи низький за наступним правилом:
- якщо рейтинг вищий або рівний 8 - то він високий (`High`)
- якщо рейтинг від 6 (включно) до 8 (не включно) - то він середній (`Medium`)
- інакше - низький (`Low`).

In [35]:
df['Rating Category'] = df.Rating.apply(lambda x: 'High' if x >= 8 else 'Medium' if 6 <= x < 8 else 'Low')
df[['Rating Category', 'Rating']]

,Rating Category,Rating
0,High,9.1
1,High,9.6
2,Medium,7.4
3,High,8.4
4,Low,5.3
...,...,...
995,Medium,6.2
996,Low,4.4
997,Medium,7.7
998,Low,4.1


Або правильніше:

In [13]:
def get_rating_category(val):
  if val >= 8:
    return "High"
  elif val >= 6 and val < 8:
    return "Medium"
  else:
    return "Low"

In [14]:
df['Rating Category'] = df.Rating.apply(get_rating_category)

3.2. Виведіть, який відсоток даних становить кожне зі значень категорій рейтингу.

In [36]:
df['Rating Category'].value_counts(normalize=True)

,proportion
Rating Category,
Medium,0.356
High,0.329
Low,0.315


4.1. Запустіть код нижче аби створити змінну `time`. Ця змінна містить час в форматі "година:хвилина". Напишіть функцію `get_hour`, яка буде зі змінної формату як `time` дістати лише години і конвертувати їх у тип `int`.

Запустіть написану функцію на змінній `time` та перевірте правильність роботи функції.

In [19]:
time = '11:28'

In [20]:
def get_hour(x):
  return int(x.split(':')[0])
get_hour(time)

11

In [15]:
# option 1

def get_hour(time):
    return int(time.split(':')[0])

# option 2
def get_hour(time):
    return int(time[:2])

4.2. З допомогою функції `get_hour` і метода `DataFrame.apply()` створіть нову колонку `Hour` в нашому наборі даних, яка буде містити години здійснення покупки, обчислені на основі часу здійснення покупки `Time`.

In [21]:
df['Hour'] = df.Time.apply(get_hour)
df['Hour']

,Hour
0,13
1,10
2,13
3,20
4,10
...,...
995,13
996,17
997,13
998,15


4.3. Виведіть кількість значень за кожною годиною покупки і дайте відповідь "О яких годинах найбільше здійснюють покупок покупці і чому на Ваш погляд саме о цих годинах? Чи це логічно, що найбільше покупок саме в ці години?".

In [22]:
df['Hour'].value_counts().sort_values()

,count
Hour,
17,74
20,75
16,77
14,83
12,89
11,90
18,93
10,101
15,102


Найбільше покупок покупці здійснюють о 19, 13, 15, 10 годинах.

Логічно, що о 19 годині здійснюється найбільше покупок, оскільки це кінець робочого дня, в людей з'являється час і потреба в закупах.

В 13 годині може бути більше покупок через обідні перерви. В людей з'являється час, бажання поїсти.

## Альтернативно:

In [23]:
df.groupby('Hour').size()

,0
Hour,
10,101
11,90
12,89
13,103
14,83
15,102
16,77
17,74
18,93


Найбільше покупок відбувається о 19:00 — 113 разів. Це може бути пов'язано з тим, що багато людей здійснюють покупки після роботи чи навчання, ввечері плануючи вечерю або інші домашні потреби. Також це може бути час, коли супермаркети пропонують спеціальні знижки або акції на кінець дня, щоб збільшити продажі.

В години обідньої перерви, о 13:00, також спостерігається велика кількість покупок — 103. Це може свідчити про те, що люди використовують свою перерву на обід, щоб зробити швидкі покупки або купити щось необхідне в середині робочого дня.

Відносно високі показники покупок о 15:00 — 102 покупки — можуть бути пов'язані з завершенням робочого дня для деяких людей, або з початком знижок та акцій, які часто стартують у другій половині дня.

Що стосується 10:00 ранку з кількістю покупок 101, це може бути пік ранкових покупок, коли люди, можливо, купують продукти на день або виконують планові покупки, оскільки багато супермаркетів починають роботу о цій годині. Це також час, коли супермаркети можуть бути менш забиті, що дозволяє швидше та комфортніше робити покупки.

5. Виведіть середнє значення чеку (`Total`) для покупок в містах, наявних в даних, за допомогою функції `groupby`.

In [46]:
df.groupby('City')['Total'].mean()

,Total
City,
Mandalay,319.872506
Naypyitaw,337.099715
Yangon,312.354031


6. Виведіть кількість покупок, середнє, мінімальне та максимальне значення для загальної суми чеку (Total) в містах, наявних в даних, за допомогою функції groupby.
Зробіть висновки з цих показників та дайте рекомендації для мережі супермаркетів.

In [52]:
df.groupby('City')['Total'].agg(['count', 'mean', 'min', 'max'])

,count,mean,min,max
City,,,,
Mandalay,332,319.872506,18.6375,1022.49
Naypyitaw,328,337.099715,10.6785,1042.65
Yangon,340,312.354031,12.6945,1039.29


**Середній чек**

   * Найвищий у Naypyitaw (≈337), трохи нижчий у Mandalay (≈320), найнижчий у Yangon (≈312).
   * Це свідчить, що покупці в Naypyitaw трохи більше витрачають за одну покупку.

**Мінімальний чек**

   * Найнижчий у Naypyitaw (≈10.68), трохи вищий у Yangon (≈12.69) і Mandalay (≈18.64).
   * Це може означати, що в Naypyitaw є більше дрібних покупок або частіше купують дешевші товари.

**Максимальний чек**

   * Всі три міста мають приблизно однаковий максимум (~1020–1040), тобто є покупки великих обсягів у всіх містах.

**Кількість покупок**

   * Найбільше у Yangon (340), найменше у Naypyitaw (328), різниця незначна.

### **Рекомендації для мережі супермаркетів:**

1. **Для Naypyitaw:**

   * Фокус на середніх та великих чеках: пропонувати комплекти або акції, які підвищують середню витрату.
   * Водночас є дрібні покупки — можна стимулювати їх через програми лояльності або купони.

2. **Для Mandalay і Yangon:**

   * Можна розвивати сегмент середніх чеків, бо максимум та мінімум схожі, а середній чек трохи нижчий.
   * Пропонувати upsell (додаткові товари до основної покупки), щоб підняти середній чек.

3. **Для всіх міст:**

   * Середня і максимальна витрата близькі, отже існує потенціал для крос-продажів і акцій “додай ще X і отримай бонус”.
   * Вивчити, які товари формують великі чеки, і робити на них акцент у маркетингових кампаніях.


# Розв'язок:

**Висновки:**

1. **Naypyitaw має найвищу середню суму чеку** (337.10), що може вказувати на більш високу купівельну спроможність або на те, що там продаються дорожчі товари. На це вказує і той факт, що тут найвищий максимальний чек.
2. **Mandalay має середню суму чеку**, схожу на Yangon (319.87), але із нижчим максимальним чеком (1022.49), що може вказувати на більш однорідні покупки.

**Рекомендації:**

1. **Naypyitaw**:
   - Оскільки Naypyitaw має найвищу середню і максимальну суму чеків, маркетингові стратегії можуть бути спрямовані на преміальний сегмент ринку. Пропозиція ексклюзивних товарів або брендів, які приваблюють клієнтів з вищою купівельною спроможністю, може ще більше підвищити продажі.
   - Запровадження спеціальних подій чи заходів, що пропонують унікальний досвід покупок, може допомогти зберегти існуючих клієнтів і залучити нових.
   - Розробка програм лояльності, які нагороджують клієнтів за частіші та більші покупки, може заохотити збільшення середнього чеку.

2. **Mandalay**:
   - З огляду на нижчий максимальний чек, супермаркет у Mandalay може зосередитись на пропозиціях, які підкреслюють цінність і доступність, щоб привабити більше клієнтів, які здійснюють регулярні покупки.
   - Проведення акцій і знижок може стимулювати клієнтів купувати більше за один візит, збільшуючи середню суму чеку.
   - Вивчення попиту на певні категорії товарів і адаптація асортименту також може допомогти підвищити суму середнього чеку.

**Додаткові рекомендації:**

- **Для обох міст**: Важливо провести детальний аналіз покупців, щоб зрозуміти їхні переваги та поведінку. Використання CRM(Customer Relationship Management)-систем може допомогти зібрати та аналізувати дані про клієнтів, щоб створювати персоналізовані пропозиції.
- **Оцінка запасів та управління ними**: Ефективне управління запасами може допомогти забезпечити наявність популярних товарів і своєчасне видалення непопулярних, щоб максимізувати прибуток і мінімізувати втрати.
- **Аналіз конкурентів**: Розуміння того, як конкуренти впливають на продажі, може допомогти визначити унікальні торгові пропозиції та виробити конкурентні стратегії ціноутворення.

7. Ми внесли деякі зміни в наш набір даних, тож, давайте запишемо результат у файл. Запишіть оновлену версію даних у файл `supermarket_sales_enriched.csv`. Зверніть увагу, аби не записати нічого зайвого, наприклад, індекси рядків нам в файлі не потрібні, оскільки вони стандартні.

In [61]:
df.to_csv('supermarket_sales_enriched.csv', index=False)
pd.read_csv('supermarket_sales_enriched.csv', nrows=5)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating,Rating Category,Hour
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,1/5/2019,13:08,Ewallet,522.83,4.761905,26.1415,9.1,High,13
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,3/8/2019,10:29,Cash,76.40,4.761905,3.8200,9.6,High,10
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,3/3/2019,13:23,Credit card,324.31,4.761905,16.2155,7.4,Medium,13
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,1/27/2019,20:33,Ewallet,465.76,4.761905,23.2880,8.4,High,20
4,373-73-7910,A,Yangon,Normal,Male,Sports and travel,86.31,7,30.2085,634.3785,2/8/2019,10:37,Ewallet,604.17,4.761905,30.2085,5.3,Low,10
